syntax = "proto3";

option java_multiple_files = true;
option java_package = "com.example.shopping";
option java_outer_classname = "ShoppingProto";

package shopping;

service Market {
    rpc RegisterSeller(RegisterSellerRequest) returns (RegisterSellerResponse);
    rpc SellItem(SellItemRequest) returns (SellItemResponse);
    rpc UpdateItem(UpdateItemRequest) returns (Response);
    rpc DeleteItem(DeleteItemRequest) returns (Response);
    rpc DisplaySellerItems(DisplaySellerItemsRequest) returns (DisplaySellerItemsResponse);
    rpc SearchItem(SearchItemRequest) returns (SearchItemResponse);
    rpc BuyItem(BuyItemRequest) returns (BuyItemResponse);
    rpc AddToWishList(AddToWishListRequest) returns (Response);
    rpc RateItem(RateItemRequest) returns (Response);
}

service NotifyClient {
    rpc Notify(Notification) returns (Response);
}

message RegisterSellerRequest {
    string uuid = 1;
    string address = 2;
}

message RegisterSellerResponse {
    string status = 1;
}

message SellItemRequest {
    string seller_uuid = 1;
    string product_name = 2;
    string category = 3;
    int32 quantity = 4;
    string description = 5;
    string seller_address = 6;
    float price = 7;
}

message SellItemResponse {
    string status = 1;
    string item_id = 2;
}

message UpdateItemRequest {
    string seller_uuid = 1;
    string item_id = 2;
    float new_price = 3;
    int32 new_quantity = 4;
    string seller_address = 5;
}

message DeleteItemRequest {
    string seller_uuid = 1;
    string item_id = 2;
    string seller_address = 3;
}

message DisplaySellerItemsRequest {
    string seller_uuid = 1;
    string seller_address = 2;
}

message DisplaySellerItemsResponse {
    repeated DisplayItem items = 1;

    message DisplayItem {
        string item_id = 1;
        float price = 2;
        string product_name = 3;
        string category = 4;
        string description = 5;
        int32 quantity = 6;
        string seller_address = 7;
        float rating = 8;
    }
}

message SearchItemRequest {
    string item_name = 1;
    string category = 2;
}

message SearchItemResponse {
    repeated SearchItemResponse.Item items = 1;

    message Item {
        string item_id = 1;
        float price = 2;
        string product_name = 3;
        string category = 4;
        string description = 5;
        int32 quantity = 6;
        float rating = 7;
        string seller_address = 8;
    }
}

message BuyItemRequest {
    string item_id = 1;
    int32 quantity = 2;
    string buyer_address = 3;
}

message BuyItemResponse {
    string status = 1;
}

message AddToWishListRequest {
    string item_id = 1;
    string buyer_address = 2;
}

message RateItemRequest {
    string item_id = 1;
    string buyer_address = 2;
    int32 rating = 3;
}

message Notification {
    string message = 1;
}

message Response {
    string status = 1;
}
